components/router/src/main.rs

```diff
@@ -0,0 +1,182 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(#400):
+// Instead of passing in a block_size, we should get this data from the backend component's config.
+// What changes need to be made:
+// 1. Take as an argument the name of the backend component.
+// 2. Update the backend component to produce a config in a standard location.
+// 3. Update the KvRouter to read the config from the backend component.
+
+use std::collections::HashMap;
+
+use clap::Parser;
+
+use dynamo_llm::kv_router::{
+    protocols::WorkerSelectionResult,
+    scheduler::{KvSchedulerError, SchedulingRequest},
+    scoring::ProcessedEndpoints,
+    KvRouter, WorkerSelector,
+};
+use dynamo_runtime::{
+    logging, pipeline::network::Ingress, DistributedRuntime, Result, Runtime, Worker,
+};
+use rand::Rng;
+
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Namespace for the distributed component
+    #[arg(long)]
+    namespace: String,
+
+    /// Component name for the service
+    #[arg(long, default_value = "kv_aware_router")]
+    component: String,
+
+    /// Block size for the router
+    #[arg(long)]
+    block_size: usize,
+}
+
+fn main() -> Result<()> {
+    logging::init();
+    let worker = Worker::from_settings()?;
+    worker.execute(app)
+}
+
+async fn app(runtime: Runtime) -> Result<()> {
+    let args = Args::parse();
+    let runtime = DistributedRuntime::from_settings(runtime).await?;
+
+    let component = runtime
+        .namespace(&args.namespace)?
+        .component(&args.component)?;
+
+    let selector = Box::new(CustomeWorkerSelector);
+
+    let router = KvRouter::new(component.clone(), args.block_size, Some(selector)).await?;
+    let router = Ingress::for_engine(router)?;
+
+    component
+        .service_builder()
+        .create()
+        .await?
+        .endpoint("generate")
+        .endpoint_builder()
+        .handler(router)
+        .start()
+        .await
+}
+
+pub struct CustomeWorkerSelector;
+
+impl WorkerSelector for CustomeWorkerSelector {
+    fn select_worker(
+        &self,
+        workers: &ProcessedEndpoints,
+        request: &SchedulingRequest,
+        kv_block_size: usize,
+    ) -> Result<WorkerSelectionResult, KvSchedulerError> {
+        let mut worker_scores = HashMap::new();
+        let mut max_waiting = 0.0;
+
+        // Calculate worker scores and find max waiting requests
+        for (i, w) in workers.endpoints.iter().enumerate() {
+            let worker_id = workers.worker_ids[i];
+
+            // Calculate score similar to Python version
+            if let Some(score) = request.overlap.scores.get(&worker_id) {
+                let score = *score as f64 * kv_block_size as f64 / request.isl_tokens as f64;
```

Comment:
is `request.isl_tokens` validated somewhere earlier to be non 0?

Reply:
it is checked, but the custody chain is broken.  good catch

Reply:
fixed